### PR TITLE
Handle single container and --rm configuration

### PR DESCRIPTION
**Pull request by pecigonzalo**
## What?

- Reworked the test to use multiple containers and fixtures as it was otherwise hard to do multiline outputs
- Included the fix from #210 for the `while` loop that otherwise prevents us from testing multiple containers and creates false positives

##### Fixes:
- #220 

##### Relates:
- #221 
- #211 
- #212 

##### Test output:
```
[1G   Logs: Detect some containers KO[K[77G1/3[2Gbats-mock(buildkite-agent): got buildkite-agent meta-data exists docker-compose-plugin-built-image-tag-myservice
bats-mock(buildkite-agent): running exit 1
bats-mock(buildkite-agent): command input is 
bats-mock(buildkite-agent): command result was 1
bats-mock(buildkite-agent): result 1
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 build --pull myservice
bats-mock(docker-compose): running echo built myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 up -d --scale myservice=0 myservice
bats-mock(docker-compose): running echo ran myservice dependencies
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 run --name buildkite1111_myservice_build_1 --rm myservice echo hello world
bats-mock(docker-compose): running echo ran myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 -q
bats-mock(docker): running cat tests/fixtures/id-multiple-services.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect -f {{if ne 0 .State.ExitCode}}{{.Name}}.{{.State.ExitCode}}{{ end }} 456456 789789
bats-mock(docker): running echo 456456.1
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 --format table {{.Label "com.docker.compose.service"}}\t{{ .ID }}\t{{ .Status }}
bats-mock(docker): running cat tests/fixtures/service-id-exit-multiple-services-failed.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 --format {{.ID}}\t{{.Label "com.docker.compose.service"}}
bats-mock(docker): running cat tests/fixtures/id-service-multiple-services.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect --format={{.State.ExitCode}} 456456
bats-mock(docker): running echo 1
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker logs --timestamps --tail 5 456456
bats-mock(docker): running exit 0
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker logs -t 456456
bats-mock(docker): running exit 0
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect --format={{.State.ExitCode}} 789789
bats-mock(docker): running echo 0
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(buildkite-agent): got buildkite-agent artifact upload docker-compose-logs/*.log
bats-mock(buildkite-agent): running exit 0
bats-mock(buildkite-agent): command input is 
bats-mock(buildkite-agent): command result was 0
bats-mock(buildkite-agent): result 0
bats-mock(buildkite-agent): unstubbing
bats-mock(docker-compose): unstubbing
bats-mock(docker): unstubbing
[1G ✓ Logs: Detect some containers KO[K

[0m[1G   Logs: Detect all containers OK[K[77G2/3[2Gbats-mock(buildkite-agent): got buildkite-agent meta-data exists docker-compose-plugin-built-image-tag-myservice
bats-mock(buildkite-agent): running exit 1
bats-mock(buildkite-agent): command input is 
bats-mock(buildkite-agent): command result was 1
bats-mock(buildkite-agent): result 1
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 build --pull myservice
bats-mock(docker-compose): running echo built myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 up -d --scale myservice=0 myservice
bats-mock(docker-compose): running echo ran myservice dependencies
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 run --name buildkite1111_myservice_build_1 --rm myservice echo hello world
bats-mock(docker-compose): running echo ran myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 -q
bats-mock(docker): running cat tests/fixtures/id-multiple-services.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect -f {{if ne 0 .State.ExitCode}}{{.Name}}.{{.State.ExitCode}}{{ end }} 456456 789789
bats-mock(docker): running echo
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 --format {{.ID}}\t{{.Label "com.docker.compose.service"}}
bats-mock(docker): running cat tests/fixtures/id-service-multiple-services.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect --format={{.State.ExitCode}} 456456
bats-mock(docker): running echo 0
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker inspect --format={{.State.ExitCode}} 789789
bats-mock(docker): running echo 0
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): unstubbing
bats-mock(docker-compose): unstubbing
bats-mock(buildkite-agent): unstubbing
[1G ✓ Logs: Detect all containers OK[K

[0m[1G   Logs: Skip output if there are no containers for a project[K[77G3/3[2Gbats-mock(buildkite-agent): got buildkite-agent meta-data exists docker-compose-plugin-built-image-tag-myservice
bats-mock(buildkite-agent): running exit 1
bats-mock(buildkite-agent): command input is 
bats-mock(buildkite-agent): command result was 1
bats-mock(buildkite-agent): result 1
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 build --pull myservice
bats-mock(docker-compose): running echo built myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 up -d --scale myservice=0 myservice
bats-mock(docker-compose): running echo ran myservice dependencies
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker-compose): got docker-compose -f docker-compose.yml -p buildkite1111 run --name buildkite1111_myservice_build_1 --rm myservice echo hello world
bats-mock(docker-compose): running echo ran myservice
bats-mock(docker-compose): command input is 
bats-mock(docker-compose): command result was 0
bats-mock(docker-compose): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 -q
bats-mock(docker): running echo
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 --format {{.ID}}\t{{.Label "com.docker.compose.service"}}
bats-mock(docker): running cat tests/fixtures/id-service-no-services.txt
bats-mock(docker): command input is 
bats-mock(docker): command result was 0
bats-mock(docker): result 0
bats-mock(docker): unstubbing
bats-mock(docker-compose): unstubbing
bats-mock(buildkite-agent): unstubbing
[1G ✓ Logs: Skip output if there are no containers for a project[K
[0m
3 tests, 0 failures
```

### CODE REVIEW

**Comment by pecigonzalo** [comment]
@ags @toolmantim  @williams4027

File: commands/run.sh [hunk 1]
```diff
@@ -275,10 +275,14 @@ if [[ -n "${BUILDKITE_AGENT_ACCESS_TOKEN:-}" ]] ; then
...
 
     # Get list of failed containers
-    failed_containers=($(
-      docker inspect -f '{{if ne 0 .State.ExitCode}}{{.Name}}.{{.State.ExitCode}}{{ end }}' \
-      $(docker_ps_by_project -q)
-    ))
+    containers=($(docker_ps_by_project -q))
+    failed_containers=()
+    if [[ 0 != "${#containers[@]}" ]] ; then
+      failed_containers=($(
```
Comment (lox) [inline commands/run.sh]: This is problematic, it's relying on implicit token splitting from the output of "$(.." into an array. 

Check out https://github.com/koalaman/shellcheck/wiki/SC2207, would suggest something like:

```
failed_containers=()
while IFS='' read -r line; do 
  array+=("$line") 
done < <(docker inspect -f '{{if ne 0 .State.ExitCode}}{{.Name}}.{{.State.ExitCode}}{{ end }}' "${containers[@]}")
```

File: commands/run.sh [hunk 2]
```diff
@@ -275,10 +275,14 @@ if [[ -n "${BUILDKITE_AGENT_ACCESS_TOKEN:-}" ]] ; then
   if [[ "$(plugin_read_config CHECK_LINKED_CONTAINERS "true")" != "false" ]] ; then
 
     # Get list of failed containers
-    failed_containers=($(
-      docker inspect -f '{{if ne 0 .State.ExitCode}}{{.Name}}.{{.State.ExitCode}}{{ end }}' \
-      $(docker_ps_by_project -q)
-    ))
+    containers=($(docker_ps_by_project -q))
```
Comment (lox) [inline commands/run.sh]: This relies on splitting too.

**Comment by lox** [comment]
Let's merge this and fix the shellcheck issues in master, thanks @pecigonzalo

File: commands/run.sh (hunk 1 above)
Comment (pecigonzalo) [reply commands/run.sh]: @lox as mentioned on #210 that breaks BATS testing, that is why I'm using the other method.

**Comment by pecigonzalo** [comment]
@lox I see the tests were fixed in https://github.com/buildkite-plugins/docker-compose-buildkite-plugin/pull/226 how did you manage to work the BATS issues that I mentioned?
Was that fixed in a new version of the BATS tested? As mentioned here, it was eating the output of `$()` invokations otherwise, making tests fail.

**Pull request merged**